CTC segmentation for Speechbrain

lumaku opened the pull request:
Text-to-audio alignment using CTC segmentation for Speechbrain.

CTC segmentation is useful to align utterances in long audio files and also to obtain a log probability estimate of how well does the ground truth fit to the audio data. The alignment requires only the CTC log posterior probabilities of a pre-trained ASR model.
https://arxiv.org/abs/2007.09127

Example:

```python
from speechbrain.pretrained import EncoderDecoderASR
from speechbrain.alignment.ctc_segmentation import CTCSegmentation

# Requires a model with CTC output
asr_model = EncoderDecoderASR.from_hparams(source="speechbrain/asr-transformer-transformerlm-librispeech")
aligner = CTCSegmentation(asr_model, kaldi_style_text=False)

# example file included in the speechbrain repository
audio_path = "./samples/audio_samples/example1.wav"
text = ["THE BIRCH CANOE", "SLID ON THE", "SMOOTH PLANKS"]
segments = aligner(audio_path, text, name="example1")
print(segments)
# example1_0000 example1 0.04 0.70 -0.0122 THE BIRCH CANOE
# example1_0001 example1 0.97 1.97 -0.0295 SLID ON THE
# example1_0002 example1 1.97 3.00 -0.0258 SMOOTH PLANKS

# here: from mini_librispeech dev data
# !wget https://www.openslr.org/resources/31/dev-clean-2.tar.gz
# !tar -xvzf dev-clean-2.tar.gz
audio_path = "./LibriSpeech/dev-clean-2/1272/135031/1272-135031-0003.flac"
text = ["THE LITTLE GIRL", "HAD BEEN ASLEEP", "BUT SHE HEARD THE RAPS", "AND OPENED THE DOOR"]
# align
segments = aligner(audio_path, text, name="1272-135031-0003")
print(segments)
# 1272-135031-0003_0000 1272-135031-0003 0.12 0.74 -0.0173 THE LITTLE GIRL
# 1272-135031-0003_0001 1272-135031-0003 0.74 1.54 -0.0012 HAD BEEN ASLEEP
# 1272-135031-0003_0002 1272-135031-0003 1.78 3.30 -0.0581 BUT SHE HEARD THE RAPS
# 1272-135031-0003_0003 1272-135031-0003 3.30 4.02 -0.0085 AND OPENED THE DOOR
```

Some notes / questions:

* This module requires the ctc-segmentation python package. I did not add ctc-segmentation to the requirements yet, as it is only needed for the alignment use-case. It can be installed via pip: `pip install ctc-segmentation` or from the Github repo https://github.com/lumaku/ctc-segmentation It is at least needed for the tests. What would be a good way to add it?

* As the ASR model in the example has very long tokens (e.g., `▁CHARACTERISTIC` is a single token), then the time resolution is bad, maybe it's only useful for sentences (instead of words). Is there by chance a pre-trained model with CTC that only contains the single characters as tokens?

* For the interface to speechbrain that obtains the CTC log probabilites, I assumed that the CTC functions of `EncoderDecoderASR` are in `<asr_module>.modules.decoder.ctc_forward_step`. How reliable is this function? If not, what is the best method to obtain these log posteriors?

* Tests: Currently, the test loads a pre-trained model. This may take too long for a test. Do you prefer a mock model?

mravanelli commented:
Hi @lumaku ,
thank you very much for this PR. It looks really like a great job! Here are my initial comments:
1. More than alignments over characters or sub-word tokens, I think it is interesting to have alignments over phonemes. This is still used for training some TTS models and many linguists use phone alignment for some studies. Did you try it with phonemes? We can share some models trained on TIMIT (and maybe train models on LibriSpeech for that). It would be interesting to have a performance comparison with the Montreal Force Alignment (based on Kaldi).  
2. Did you assess the quality of the alignment using some of our SpeechBrain models?
3. As for showing the alignment did you consider the praat format as well? Many people would like to visualize the alignment with it
4. As for the dependency with `ctc-segmentation`, I think it is fine to keep it as an optional dependency.  I suggest adding in the code a check that asks users to install it if not present in the system. As an alternative, you can import all your code in SpeechBrain with proper mention of your paper and the original repository (but that's up to you).
5.  It looks like, beyond `ctc-segmentation`, there is another external dependency called `typeguard`. Do you think we can avoid it (or including it into CTC-segmentation)?  SpeechBrain tries to minimize as much as possible the use of external dependencies.
6. `<asr_module>.modules.decoder.ctc_forward_step` => I think this is fine. Maybe we can do a check and raise an error if this module (and the other things requires by the library) are not present.
7. Some docstrings are not compliant with the SpeechBrain standard. 
8. I tried to run the code snippet and everything seems to work fine:

```
example1_0000 example1 0.04 0.70 -0.0304 THE BIRCH CANOE
example1_0001 example1 0.97 1.97 -0.0131 SLID ON THE
example1_0002 example1 1.97 3.00 -0.0375 SMOOTH PLANKS
```

As for the scoring, I got different numbers than what you achieved. I guess they should be the same, right?

9. To make users aware of this new functionality we might wanna think about a GitHub gist or even a tutorial. But first of all, it would be better to assess the performance (e.g, on phoneme alignment on TIMIT or LibriSpeech).

lumaku commented:
1. Alignment over phonemes:
I haven't tested it, but it should work. Of course, it needs a function that helps to convert text to phonemes and a compatible ASR model...
In the paper, we only compared it with the alignments from Gentle, but not with Montreal Force Alignment. I also think that the use-case is different? We programmed the algorithm with the main purpose to find begin and endings of utterances, i.e. it will omit the space character if text is splitted as words. The algorithm can skip segments and filter out bad parts (So, not really forced alignment). We used CTC segmentation on audio files of up to 3h of audio.
I think TTS also requires the duration of phonemes? CTC activations spike when a token "occurs" - which is contrary to hybrid DNN/HMM systems that indirectly infer phoneme duration by classifying frames. Nevertheless, some information of token timing is preserved, from the example:

```python
for i, j in zip(segments.ground_truth_mat, segments.timings):
    if i[0] <= 0:
        continue
    token = segments.config.char_list[int(i[0])]
    print(f"{j:.3f} {token}")
# 0.320 ▁THE
# 0.479 ▁LITTLE
# 0.719 ▁GIRL
# 0.999 ▁HAD
# 1.159 ▁BEEN
# 1.279 ▁A
# 1.518 SLEEP
# 2.278 ▁BUT
# 2.397 ▁SHE
# 2.677 ▁HE
# 2.717 ARD
# 2.877 ▁THE
# 2.997 ▁RA
# 3.197 P
# 3.277 S
# 3.356 ▁AND
# 3.716 ▁OPEN
# 3.756 ED
# 3.876 ▁THE
# 3.996 ▁DOOR
```
From my experience, the CTC activations are sometimes shifted by a few frames, sometimes up to 300 ms! Phonetic information about characters is not equally distributed.


2. Tests with Speechbrain models:
In general, alignment performance depends on a few factors: Time resolution, token dictionary, how well data fits to ASR model, and also on the data. I only tested the `speechbrain/asr-transformer-transformerlm-librispeech` model for my initial implementation. From the confidence score, it seems that the model fits very well to the data and is pretty good. The time resolution is a little bit lower, about 40 ms, while the models I previously worked with had 32 ms (638 vs 512 audio sample points per CTC output index @ 16kHz). Due to the 5000 tokens of this model that sometimes consisted of full words, the estimated alignments were inaccurate.
A single-character model would achieve a better accuracy. It would make sense to train such a model. Preferably with lower subsampling; if available RNN-based models are more memory efficient on longer audio files. In my experiments with espnet, Sinc-DSConv-BLSTMP models were fast and memory-efficient, but that's maybe another pull request......


3. Praat format:
I haven't thought about this yet. I'm not sure what information is needed for this format. Is there somewhere a good helper function to write this format?


4. `ctc-segmentation` as dependency:
As the module needs cython to compile the code, it's probably better to keep it out of Speechbrain... 
If you have improvements to the algorithm or some want to test other algorithms, I also welcome pull requests in the ctc segmentation repository.
I'll add a short check in the code as you proposed, and I would add it in `setup.py` as `extras_require` for the tests, as in setuptools documentation:
https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#optional-dependencies

5. `typeguard` is not needed. I'll remove it
6. `<asr_module>.modules.decoder.ctc_forward_step` --> OK!
7. docstrings: I'll fix that in the next commit.
8. Your output is correct - in the example, I used slightly different parameters....
9. Regarding demo: I'll write a short gist. Maybe with a short multiprocessing demo and some downloaded files. That would be a good starting point for a comparison....

mravanelli commented:
Hi @lumaku!
thank you for the updates. I think we can easily train a simple model with TIMIT that predicts phoneme as targets without downsampling. It could be interesting to see how `ctc-segmentation` works on that compared to other HMM-based methods we have in speechbrain for alignment.  Let's see if the label delay happens in this context as well.
Praat is often used to visualize alignment. They have a specific format for that. As far as I remember there is also some python library to read/save praat alignments (but this unfortunately it requires an extra dependency).

lingjzhu commented:
> 1. Alignment over phonemes:
>    I haven't tested it, but it should work. Of course, it needs a function that helps to convert text to phonemes and a compatible ASR model...
>    In the paper, we only compared it with the alignments from Gentle, but not with Montreal Force Alignment. I also think that the use-case is different? We programmed the algorithm with the main purpose to find begin and endings of utterances, i.e. it will omit the space character if text is splitted as words. The algorithm can skip segments and filter out bad parts (So, not really forced alignment). We used CTC segmentation on audio files of up to 3h of audio.
>    I think TTS also requires the duration of phonemes? CTC activations spike when a token "occurs" - which is contrary to hybrid DNN/HMM systems that indirectly infer phoneme duration by classifying frames. Nevertheless, some information of token timing is preserved, from the example:


Thanks for this helpful discussion! I am currently working on phoneme alignments by training transformer-based phoneme recognition models for different languages (English & Mandarin). Getting phoneme duration is fundamental for many speech research projects. 

I also find that CTC spikes des not directly classify frames, which might not be optimal for getting duration. Do you think DNN/HMM systems are better suited for this tasks?

mravanelli commented:
It could be interesting a comparison between CTC and HMM-DNN for phoneme alignment.  Note that in SpeechBrain we already have some recipes for doing alignment with HMM/DNN (see https://github.com/speechbrain/speechbrain/tree/develop/recipes/TIMIT/Alignment). The model uses a simple neural network trained with a special cost function that embeds HMMs with the forward or Viterbi algorithm. Feel free to play with that. On TIMIT it works well, but as far as I remember it was hard to get good results on long sentences.

In general, I think it could be great to provide to the users a way to derive high-quality alignments over phonemes (similar to the Montreal force alignment based on Kaldi)

mravanelli commented:
Hi @lumaku,any news on that?

lumaku commented:
Sorry for the delay.

* Code is complete from my side and the test issues should be fixed
* About the unit test. In the test, a pre-trained model is automatically downloaded, is this OK or does it need a mock model for faster testing? I did not yet find a good method to set up a speechbrain mock model, if you have a function for that, let me know!
* As we discussed before, the model used in the example has a relatively low accuracy, we should change that as soon there is a pre-trained char-based model available...

lumaku commented:
@lingjzhu 
> I also find that CTC spikes des not directly classify frames, which might not be optimal for getting duration. Do you think DNN/HMM systems are better suited for this task?

Hybrid HMM/DNN systems inherently detect phoneme duration because of their different HMM model. In comparison, CTC inserts a "blank" between token states and the network is not trained with time-synchronous labels. A tied triphone model does not have those blank states.

There are other suitable HMM structures for end-to-end ASR, maybe the ASG criterion, or maybe by applying the classic triphone HMM structure using a WFST (facebooks GTN, or K2)?

In comparison, CTC segmentation estimates utterance start and end by looking at the token timings: The utterance "border" is just in the middle between two tokens. From the example above, if we want to estimate the duration of the word "GIRL", we would first estimate the middle between ▁LITTLE to ▁GIRL (0.599); and ▁GIRL to ▁HAD (0.859); and then take the difference (~300 ms). In this case, it is not very accurate, but with char-based models, this estimation from the CTC activations was surprisingly close to hand-crafted labels in our evaluation (on tedlium2, librivox).

lumaku commented:
@mravanelli Could you approve the workflow so that I can see whether it runs through (it runs on my computer...), or if there are still some unforeseen issues?

lumaku commented:
So, from my side, the tests ran through and this would be OK to merge. We can change to another example ASR model as soon as a better model is available, and I'll provide an example gist some time later...
Let me know if this is OK for you and if you require some further changes.

TParcollet commented on setup.py:
```diff
@@ -31,6 +31,7 @@
         "tqdm",
         "huggingface_hub",
     ],
+    extras_require={"full": ["ctc-segmentation<1.8,>=1.6.6"],},  # noqa: E231
```
I don't think we want that from scratch. Please check instead if it is installed, and prompt the users to do so if not (directly on the ctc_segmentation.py file). + Why do you need it, can't we code our own segmenter?

lumaku replied:
I put this as a separate dependency, and didn't integrate the full code, because: On install, the CTC segmentation code needs to be cythonized (to be faster), and therefore requires a separate installation step (or, dynamic compilation with `pyximport`). I think that it is easier to install it from a Python package. See point 4 of the discussion above.

That said, I personally see also other advantages for using a package: Any change is also immediately available in other ASR frameworks (e.g., for testing). I highly welcome pull requests that improve or extend the code. Also, I prefer to maintain only one version of the code...

If you want to include the full code in speechbrain -
This includes `ctc_segmentation.py` and `ctc_segmentation_dyn.pyx` from:
https://github.com/lumaku/ctc-segmentation/tree/master/ctc_segmentation

So, because it is easier to install from Python package, I integrated it as an optional dependency.
For packages that are installed with setuptools, those optional requirements usually are defined in `extras_require`
as described in https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#optional-dependencies
Optional dependencies can then be installed together with the package by writing the dependency in brackets:
`pip install speechbrain[full]`.

lumaku replied:
@TParcollet I was not entirely sure how you prefer to handle the optional dependency on `ctc-segmentation`.
Can you check if the current code is OK for you?

I additionally made a small change to how timing is calculated: I originally programmed this for espnet, that discards trailing audio input data; speechbrain frontend is a little bit different. I simply changed the calculation from integer to floating point. It's only a few ms, but it seems to be more accurate... If you retry with your previous short test audio file, maybe the alignments are better...?

TParcollet commented on speechbrain/alignment/ctc_segmentation.py:
```diff
@@ -0,0 +1,652 @@
+#!/usr/bin/env python3
+# 2021, Technische Universität München, Ludwig Kürzinger
+"""Perform CTC segmentation to align utterances within audio files."""
```
Mention here that one package needs to be installed.

lumaku replied:
OK - will be included in the next commit

TParcollet commented on speechbrain/alignment/ctc_segmentation.py:
```diff
@@ -0,0 +1,652 @@
+    The human-readable output can be configured with
+    the printing options.
+
+    Properties
```
Properties should be changed to Arguments. Please see: https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/nnet/CNN.py For the formalism. You might need to specify the type of each argument as well.

lumaku replied:
This object is automatically generated and just a container for results, including a function for pretty printing.
There is no interface for arguments needed - I'll remove it.

TParcollet commented on speechbrain/alignment/ctc_segmentation.py:
```diff
@@ -0,0 +1,652 @@
+            the parameter ``samples_to_frames_ratio``. Recommended for
+            longer audio files: "auto".
+        **ctc_segmentation_args
+            Parameters for CTC segmentation.
```
Referring to what? Could you please mention the Class or where the user look at if they want to know what are the possible arguments?

lumaku replied:
I'll refer to the corresponding function.

TParcollet commented on speechbrain/alignment/ctc_segmentation.py:
```diff
@@ -0,0 +1,652 @@
+    def set_config(self, **kwargs):
+        """Set CTC segmentation parameters.
+
+        Parameters for timing
```
Same -> Arguments

lumaku replied:
OK!

TParcollet reviewed (changes requested):
Thank you so much for this awesome work! I will test it. Could you please fix the different comments?

TParcollet commented:
I tested with a custom french audio file, here is the output (and with our wav2vec2 french model) 
<img width="916" alt="Capture d’écran 2021-07-07 à 11 21 13" src="https://user-images.githubusercontent.com/11910731/124735087-002a9f80-df16-11eb-899d-476db3f0c499.png">

It looks like the beginning of each segment is slightly too soon -> I have the end of the previous one. But it's really slight. I wonder if this is due to the model used ... 

@lumaku I noticed that the param vocab_size  was needed in the hyperparams file. This isn't usual. Maybe it should be mention somewhere, or just for it to exist. If it doesn't print something that explicit the fact that it corresponds to the vocab_size including eos, bos blank etc.

lumaku commented on .github/workflows/pythonapp.yml:
```diff
@@ -43,7 +43,7 @@ jobs:
               run: |
                   pip install -r requirements.txt
-                  pip install --editable .
+                  pip install --editable .[full]
```
This change depends on the `extras_require`, see the comments in `setup.py`.
For all tests to pass, the `ctc-segmentation` package is needed, and this change causes the installation to additionally install the `extras_require`.

lumaku commented:
Regarding `asr_model.hparams.vocab_size`:
It's only needed to get the token dictionary.
I changed this to `asr_model.tokenizer.vocab_size()`
Or, is there a better method to get the token dictionary?

Regarding timings:
I tested the code with only two pretrained Speechbrain models, and I did not get good alignments yet... It could be due to lower accuracy with larger tokens, or due to a lower time resolution. The best accuracy is with models that only have the basic characters as tokens. To read time resolution after aligning: `segments.config.index_duration_in_seconds`.

The aligner can't see the audio data, and timings are determined only from the CTC output indices. This depends on the mapping of input audio samples to CTC output. Usually, this should be OK, but I am not familiar with how the frontend in Speechbrain handles input data, e.g. is some data discarded, or is there padding?

Aligning over longer files helps to find whether it is a systematic error or due to inaccuracies; I'd like to try with `speechbrain/asr-wav2vec2-commonvoice-fr` from the hub, but it seems that this model does not have a trained CTC layer...?

In the speechbrain huggingface hub, the Chinese model is more accurate than the English example model;
with a wav from the AISHELL samples:
```python
asr_model.modules.decoder.ctc_weight = 0.5

asr_model = EncoderDecoderASR.from_hparams(source="speechbrain/asr-wav2vec2-transformer-aishell")
aligner = CTCSegmentation(asr_model, kaldi_style_text=False
text = ["住房用地供应","和管理情况以及税收政策执行", "和征管情况"]
wav = "AISHELL-1_sample/S0150/S0150_mic/BAC009S0150W0240.wav"
segments = aligner(wav, text)
print(segments)
# utt_0000 utt 0.26 2.55 -0.6748 住房用地供应
# utt_0001 utt 2.55 6.67 -0.9303 和管理情况以及税收政策执行
# utt_0002 utt 6.67 8.53 -0.5613 和征管情况

# - get time resolution:
print(segments.config.index_duration_in_seconds)
# 0.020011077680525166
```

lumaku commented:
So, I tested this with an audio file of about 8 minutes on the French model. The alignments seem to be OK. It could be that the timing information is not accurately estimated on short audio files, I'm not sure. Memory consumption at inference seemed to be a bit high.

<details><summary>+ Code</summary><div>

With `lesmiserables_t1_01_hugo.mp3` downloaded from https://archive.org/details/les_miserables_tome_1_di_0910_librivox
```python

audio = "/xxx/lesmiserables_t1_01_hugo.mp3"
text= """
Tome Un FANTINE.
Livre premier. Un juste.
Chapitre Un.
Monsieur Myriel.

En 1815, M. Charles-François-Bienvenu Myriel était évêque de Digne.
C'était un vieillard d'environ soixante-quinze ans; il occupait le siège
de Digne depuis 1806.

Quoique ce détail ne touche en aucune manière au fond même de ce que
nous avons à raconter, il n'est peut-être pas inutile, ne fût-ce que
pour être exact en tout, d'indiquer ici les bruits et les propos qui
avaient couru sur son compte au moment où il était arrivé dans le
diocèse. Vrai ou faux, ce qu'on dit des hommes tient souvent autant de
place dans leur vie et surtout dans leur destinée que ce qu'ils font. M.
Myriel était fils d'un conseiller au parlement d'Aix; noblesse de robe.
On contait de lui que son père, le réservant pour hériter de sa charge,
l'avait marié de fort bonne heure, à dix-huit ou vingt ans, suivant un
usage assez répandu dans les familles parlementaires. Charles Myriel,
nonobstant ce mariage, avait, disait-on, beaucoup fait parler de lui. Il
était bien fait de sa personne, quoique d'assez petite taille, élégant,
gracieux, spirituel; toute la première partie de sa vie avait été donnée
au monde et aux galanteries. La révolution survint, les événements se
précipitèrent, les familles parlementaires décimées, chassées, traquées,
se dispersèrent. M. Charles Myriel, dès les premiers jours de la
révolution, émigra en Italie. Sa femme y mourut d'une maladie de
poitrine dont elle était atteinte depuis longtemps. Ils n'avaient point
d'enfants. Que se passa-t-il ensuite dans la destinée de M. Myriel?
L'écroulement de l'ancienne société française, la chute de sa propre
famille, les tragiques spectacles de 93, plus effrayants encore
peut-être pour les émigrés qui les voyaient de loin avec le
grossissement de l'épouvante, firent-ils germer en lui des idées de
renoncement et de solitude? Fut-il, au milieu d'une de ces distractions
et de ces affections qui occupaient sa vie, subitement atteint d'un de
ces coups mystérieux et terribles qui viennent quelquefois renverser, en
le frappant au coeur, l'homme que les catastrophes publiques
n'ébranleraient pas en le frappant dans son existence et dans sa
fortune? Nul n'aurait pu le dire; tout ce qu'on savait, c'est que,
lorsqu'il revint d'Italie, il était prêtre.

En 1804, M. Myriel était curé de Brignolles. Il était déjà vieux, et
vivait dans une retraite profonde.

Vers l'époque du couronnement, une petite affaire de sa cure, on ne sait
plus trop quoi, l'amena à Paris. Entre autres personnes puissantes, il
alla solliciter pour ses paroissiens M. le cardinal Fesch. Un jour que
l'empereur était venu faire visite à son oncle, le digne curé, qui
attendait dans l'antichambre, se trouva sur le passage de sa majesté.
Napoléon, se voyant regardé avec une certaine curiosité par ce
vieillard, se retourna, et dit brusquement:

--Quel est ce bonhomme qui me regarde?

--Sire, dit M. Myriel, vous regardez un bonhomme, et moi je regarde un
grand homme. Chacun de nous peut profiter.

L'empereur, le soir même, demanda au cardinal le nom de ce curé, et
quelque temps après M. Myriel fut tout surpris d'apprendre qu'il était
nommé évêque de Digne.

Qu'y avait-il de vrai, du reste, dans les récits qu'on faisait sur la
première partie de la vie de M. Myriel? Personne ne le savait. Peu de
familles avaient connu la famille Myriel avant la révolution.

M. Myriel devait subir le sort de tout nouveau venu dans une petite
ville où il y a beaucoup de bouches qui parlent et fort peu de têtes qui
pensent. Il devait le subir, quoiqu'il fût évêque et parce qu'il était
évêque. Mais, après tout, les propos auxquels on mêlait son nom
n'étaient peut-être que des propos; du bruit, des mots, des paroles;
moins que des paroles, des _palabres_, comme dit l'énergique langue du
midi.

Quoi qu'il en fût, après neuf ans d'épiscopat et de résidence à Digne,
tous ces racontages, sujets de conversation qui occupent dans le premier
moment les petites villes et les petites gens, étaient tombés dans un
oubli profond. Personne n'eût osé en parler, personne n'eût même osé
s'en souvenir.

M. Myriel était arrivé à Digne accompagné d'une vieille fille,
mademoiselle Baptistine, qui était sa soeur et qui avait dix ans de
moins que lui.

Ils avaient pour tout domestique une servante du même âge que
mademoiselle Baptistine, et appelée madame Magloire, laquelle, après
avoir été _la servante de M. le Curé_, prenait maintenant le double
titre de femme de chambre de mademoiselle et femme de charge de
monseigneur.

Mademoiselle Baptistine était une personne longue, pâle, mince, douce;
elle réalisait l'idéal de ce qu'exprime le mot «respectable»; car il
semble qu'il soit nécessaire qu'une femme soit mère pour être vénérable.
Elle n'avait jamais été jolie; toute sa vie, qui n'avait été qu'une
suite de saintes oeuvres, avait fini par mettre sur elle une sorte de
blancheur et de clarté; et, en vieillissant, elle avait gagné ce qu'on
pourrait appeler la beauté de la bonté. Ce qui avait été de la maigreur
dans sa jeunesse était devenu, dans sa maturité, de la transparence; et
cette diaphanéité laissait voir l'ange. C'était une âme plus encore que
ce n'était une vierge. Sa personne semblait faite d'ombre; à peine assez
de corps pour qu'il y eût là un sexe; un peu de matière contenant une
lueur; de grands yeux toujours baissés; un prétexte pour qu'une âme
reste sur la terre.

Madame Magloire était une petite vieille, blanche, grasse, replète,
affairée, toujours haletante, à cause de son activité d'abord, ensuite à
cause d'un asthme.

À son arrivée, on installa M. Myriel en son palais épiscopal avec les
honneurs voulus par les décrets impériaux qui classent l'évêque
immédiatement après le maréchal de camp. Le maire et le président lui
firent la première visite, et lui de son côté fit la première visite au
général et au préfet.

L'installation terminée, la ville attendit son évêque à l'oeuvre"""
# text preprocessing
text = text.replace("M.", "Monsieur")
text = text.replace("1815", "dix-huit cent quinze")
text = text.replace("1806", "dix-huit cent six")
text = text.replace("?", ".").replace(";", ".").replace(",", ".").replace("--", " ")
text = " ".join(text.split())
text = text.replace(". ", ".").upper().split(".")
# alignment
asr_model = EncoderDecoderASR.from_hparams(source="speechbrain/asr-wav2vec2-commonvoice-fr")
asr_model.modules.decoder.ctc_weight = 0.5
aligner = CTCSegmentation(asr_model, kaldi_style_text=False, gratis_blank=True)
segments = aligner(audio, text)
```

</div></details>

<details><summary>+ Segments</summary><div>

You may notice that the segments start only after 28s, that's because in the first part, the reader says book title, that's it's part of librivox, ...

```
utt_0000 utt 28.08 30.32 -0.5524 TOME UN FANTINE
utt_0001 utt 31.08 32.45 -0.1081 LIVRE PREMIER
utt_0002 utt 32.96 34.16 -0.1301 UN JUSTE
utt_0003 utt 34.70 35.98 -0.3609 CHAPITRE UN
utt_0004 utt 36.76 38.08 -0.3743 MONSIEUR MYRIEL
utt_0005 utt 39.98 41.61 -1.5222 EN DIX-HUIT CENT QUINZE
utt_0006 utt 41.61 45.89 -1.2631 MONSIEUR CHARLES-FRANÇOIS-BIENVENU MYRIEL ÉTAIT ÉVÊQUE DE DIGNE
utt_0007 utt 46.64 49.23 -0.7951 C'ÉTAIT UN VIEILLARD D'ENVIRON SOIXANTE-QUINZE ANS
utt_0008 utt 49.46 53.58 -1.0346 IL OCCUPAIT LE SIÈGE DE DIGNE DEPUIS DIX-HUIT CENT SIX
utt_0009 utt 54.50 59.75 -0.8542 QUOIQUE CE DÉTAIL NE TOUCHE EN AUCUNE MANIÈRE AU FOND MÊME DE CE QUE NOUS AVONS À RACONTER
utt_0010 utt 60.10 62.09 -0.5867 IL N'EST PEUT-ÊTRE PAS INUTILE
utt_0011 utt 62.22 64.55 -0.6493 NE FÛT-CE QUE POUR ÊTRE EXACT EN TOUT
utt_0012 utt 64.86 72.81 -0.4257 D'INDIQUER ICI LES BRUITS ET LES PROPOS QUI AVAIENT COURU SUR SON COMPTE AU MOMENT OÙ IL ÉTAIT ARRIVÉ DANS LE DIOCÈSE
utt_0013 utt 73.48 74.55 -0.2002 VRAI OU FAUX
utt_0014 utt 74.82 82.31 -0.5089 CE QU'ON DIT DES HOMMES TIENT SOUVENT AUTANT DE PLACE DANS LEUR VIE ET SURTOUT DANS LEUR DESTINÉE QUE CE QU'ILS FONT
utt_0015 utt 83.50 86.94 -0.9711 MONSIEUR MYRIEL ÉTAIT FILS D'UN CONSEILLER AU PARLEMENT D'AIX
utt_0016 utt 87.22 88.73 -0.3234 NOBLESSE DE ROBE
utt_0017 utt 89.82 91.87 -0.2654 ON CONTAIT DE LUI QUE SON PÈRE
utt_0018 utt 92.06 94.53 -0.3102 LE RÉSERVANT POUR HÉRITER DE SA CHARGE
utt_0019 utt 94.86 96.99 -0.9107 L'AVAIT MARIÉ DE FORT BONNE HEURE
utt_0020 utt 97.30 99.01 -1.0155 À DIX-HUIT OU VINGT ANS
utt_0021 utt 99.32 103.37 -0.3329 SUIVANT UN USAGE ASSEZ RÉPANDU DANS LES FAMILLES PARLEMENTAIRES
utt_0022 utt 104.20 105.46 -0.4780 CHARLES MYRIEL
utt_0023 utt 105.46 107.18 -1.1617 NONOBSTANT CE MARIAGE
utt_0024 utt 107.76 108.33 -0.0567 AVAIT
utt_0025 utt 108.33 109.05 -0.0309 DISAIT-ON
utt_0026 utt 109.24 110.81 -0.0517 BEAUCOUP FAIT PARLER DE LUI
utt_0027 utt 112.00 114.06 -0.1165 IL ÉTAIT BIEN FAIT DE SA PERSONNE
utt_0028 utt 114.18 115.95 -0.8009 QUOIQUE D'ASSEZ PETITE TAILLE
utt_0029 utt 116.32 117.18 -0.4313 ÉLÉGANT
utt_0030 utt 117.48 118.51 -0.6122 GRACIEUX
utt_0031 utt 118.64 119.71 -0.2498 SPIRITUEL
utt_0032 utt 120.66 125.85 -0.3959 TOUTE LA PREMIÈRE PARTIE DE SA VIE AVAIT ÉTÉ DONNÉE AU MONDE ET AUX GALANTERIES
utt_0033 utt 127.00 128.83 -0.4148 LA RÉVOLUTION SURVINT
utt_0034 utt 129.16 131.15 -0.6118 LES ÉVÉNEMENTS SE PRÉCIPITÈRENT
utt_0035 utt 131.40 133.83 -0.8563 LES FAMILLES PARLEMENTAIRES DÉCIMÉES
utt_0036 utt 133.83 134.75 -0.1775 CHASSÉES
utt_0037 utt 134.86 135.69 -0.2947 TRAQUÉES
utt_0038 utt 135.82 137.11 -0.2428 SE DISPERSÈRENT
utt_0039 utt 138.80 140.38 -0.3732 MONSIEUR CHARLES MYRIEL
utt_0040 utt 140.40 142.37 -0.4182 DÈS LES PREMIERS JOURS DE LA RÉVOLUTION
utt_0041 utt 142.88 144.35 -0.5594 ÉMIGRA EN ITALIE
utt_0042 utt 145.15 150.06 -0.3402 SA FEMME Y MOURUT D'UNE MALADIE DE POITRINE DONT ELLE ÉTAIT ATTEINTE DEPUIS LONGTEMPS
utt_0043 utt 150.55 152.04 -0.9263 ILS N'AVAIENT POINT D'ENFANTS
utt_0044 utt 152.55 156.22 -0.9884 QUE SE PASSA-T-IL ENSUITE DANS LA DESTINÉE DE MONSIEUR MYRIEL
utt_0045 utt 157.31 160.12 -0.3579 L'ÉCROULEMENT DE L'ANCIENNE SOCIÉTÉ FRANÇAISE
utt_0046 utt 160.37 162.52 -0.6507 LA CHUTE DE SA PROPRE FAMILLE
utt_0047 utt 162.77 165.66 -0.4019 LES TRAGIQUES SPECTACLES DE 93
utt_0048 utt 166.09 172.50 -0.7227 PLUS EFFRAYANTS ENCORE PEUT-ÊTRE POUR LES ÉMIGRÉS QUI LES VOYAIENT DE LOIN AVEC LE GROSSISSEMENT DE L'ÉPOUVANTE
utt_0049 utt 172.91 177.09 -1.6390 FIRENT-ILS GERMER EN LUI DES IDÉES DE RENONCEMENT ET DE SOLITUDE
utt_0050 utt 177.91 178.68 -0.3153 FUT-IL
utt_0051 utt 178.68 183.02 -0.7917 AU MILIEU D'UNE DE CES DISTRACTIONS ET DE CES AFFECTIONS QUI OCCUPAIENT SA VIE
utt_0052 utt 183.23 189.56 -1.3320 SUBITEMENT ATTEINT D'UN DE CES COUPS MYSTÉRIEUX ET TERRIBLES QUI VIENNENT QUELQUEFOIS RENVERSER
utt_0053 utt 189.59 191.00 -0.8259 EN LE FRAPPANT AU COEUR
utt_0054 utt 191.33 198.15 -0.6863 L'HOMME QUE LES CATASTROPHES PUBLIQUES N'ÉBRANLERAIENT PAS EN LE FRAPPANT DANS SON EXISTENCE ET DANS SA FORTUNE
utt_0055 utt 198.83 200.30 -0.2961 NUL N'AURAIT PU LE DIRE
utt_0056 utt 200.99 202.28 -0.1125 TOUT CE QU'ON SAVAIT
utt_0057 utt 202.33 203.00 -0.2763 C'EST QUE
utt_0058 utt 203.00 204.34 -0.6415 LORSQU'IL REVINT D'ITALIE
utt_0059 utt 204.34 205.50 -0.2598 IL ÉTAIT PRÊTRE
utt_0060 utt 206.25 206.90 -0.1614 EN 1804
utt_0061 utt 207.97 210.28 -0.4308 MONSIEUR MYRIEL ÉTAIT CURÉ DE BRIGNOLLES
utt_0062 utt 210.69 212.09 -0.2811 IL ÉTAIT DÉJÀ VIEUX
utt_0063 utt 212.09 214.58 -0.3625 ET VIVAIT DANS UNE RETRAITE PROFONDE
utt_0064 utt 215.57 217.34 -0.1666 VERS L'ÉPOQUE DU COURONNEMENT
utt_0065 utt 217.34 219.16 -0.3601 UNE PETITE AFFAIRE DE SA CURE
utt_0066 utt 219.45 220.82 -0.0968 ON NE SAIT PLUS TROP QUOI
utt_0067 utt 221.09 222.41 -0.3949 L'AMENA À PARIS
utt_0068 utt 223.13 225.04 -0.4874 ENTRE AUTRES PERSONNES PUISSANTES
utt_0069 utt 225.31 229.80 -0.5230 IL ALLA SOLLICITER POUR SES PAROISSIENS MONSIEUR LE CARDINAL FESCH
utt_0070 utt 230.61 234.30 -0.5074 UN JOUR QUE L'EMPEREUR ÉTAIT VENU FAIRE VISITE À SON ONCLE
utt_0071 utt 234.57 235.83 -0.3311 LE DIGNE CURÉ
utt_0072 utt 235.87 237.66 -0.3142 QUI ATTENDAIT DANS L'ANTICHAMBRE
utt_0073 utt 237.87 240.58 -0.3032 SE TROUVA SUR LE PASSAGE DE SA MAJESTÉ
utt_0074 utt 241.31 242.22 -0.3510 NAPOLÉON
utt_0075 utt 242.45 246.18 -0.2590 SE VOYANT REGARDÉ AVEC UNE CERTAINE CURIOSITÉ PAR CE VIEILLARD
utt_0076 utt 246.55 247.51 -0.3290 SE RETOURNA
utt_0077 utt 247.51 251.43 -1.9178 ET DIT BRUSQUEMENT: QUEL EST CE BONHOMME QUI ME REGARDE
utt_0078 utt 252.29 252.84 -0.5249 SIRE
utt_0079 utt 253.01 254.30 -0.3924 DIT MONSIEUR MYRIEL
utt_0080 utt 254.69 256.46 -0.5346 VOUS REGARDEZ UN BONHOMME
utt_0081 utt 256.75 258.86 -0.4982 ET MOI JE REGARDE UN GRAND HOMME
utt_0082 utt 259.23 261.37 -0.2903 CHACUN DE NOUS PEUT PROFITER
utt_0083 utt 262.63 263.54 -0.4029 L'EMPEREUR
utt_0084 utt 263.54 264.50 -0.2678 LE SOIR MÊME
utt_0085 utt 264.81 267.58 -0.5795 DEMANDA AU CARDINAL LE NOM DE CE CURÉ
utt_0086 utt 267.69 273.51 -0.6112 ET QUELQUE TEMPS APRÈS MONSIEUR MYRIEL FUT TOUT SURPRIS D'APPRENDRE QU'IL ÉTAIT NOMMÉ ÉVÊQUE DE DIGNE
utt_0087 utt 275.59 277.07 -0.9276 QU'Y AVAIT-IL DE VRAI
utt_0088 utt 277.07 277.80 -0.2477 DU RESTE
utt_0089 utt 277.95 282.19 -0.9001 DANS LES RÉCITS QU'ON FAISAIT SUR LA PREMIÈRE PARTIE DE LA VIE DE MONSIEUR MYRIEL
utt_0090 utt 282.77 284.34 -0.0757 PERSONNE NE LE SAVAIT
utt_0091 utt 284.81 288.32 -1.0121 PEU DE FAMILLES AVAIENT CONNU LA FAMILLE MYRIEL AVANT LA RÉVOLUTION
utt_0092 utt 289.27 298.66 -0.7310 MONSIEUR MYRIEL DEVAIT SUBIR LE SORT DE TOUT NOUVEAU VENU DANS UNE PETITE VILLE OÙ IL Y A BEAUCOUP DE BOUCHES QUI PARLENT ET FORT PEU DE TÊTES QUI PENSENT
utt_0093 utt 299.41 300.72 -0.0832 IL DEVAIT LE SUBIR
utt_0094 utt 300.72 303.62 -0.2762 QUOIQU'IL FÛT ÉVÊQUE ET PARCE QU'IL ÉTAIT ÉVÊQUE
utt_0095 utt 304.91 305.42 -0.0168 MAIS
utt_0096 utt 305.45 306.36 -0.1922 APRÈS TOUT
utt_0097 utt 306.47 311.39 -0.8061 LES PROPOS AUXQUELS ON MÊLAIT SON NOM N'ÉTAIENT PEUT-ÊTRE QUE DES PROPOS
utt_0098 utt 311.75 312.61 -0.2856 DU BRUIT
utt_0099 utt 312.75 313.52 -0.1589 DES MOTS
utt_0100 utt 313.59 314.54 -0.0826 DES PAROLES
utt_0101 utt 314.95 316.22 -0.0909 MOINS QUE DES PAROLES
utt_0102 utt 316.25 317.44 -0.4752 DES _PALABRES_
utt_0103 utt 317.47 319.87 -0.1557 COMME DIT L'ÉNERGIQUE LANGUE DU MIDI
utt_0104 utt 320.77 321.96 -0.4920 QUOI QU'IL EN FÛT
utt_0105 utt 322.07 325.69 -1.5119 APRÈS NEUF ANS D'ÉPISCOPAT ET DE RÉSIDENCE À DIGNE
utt_0106 utt 325.97 327.46 -0.1400 TOUS CES RACONTAGES
utt_0107 utt 327.55 333.60 -0.8017 SUJETS DE CONVERSATION QUI OCCUPENT DANS LE PREMIER MOMENT LES PETITES VILLES ET LES PETITES GENS
utt_0108 utt 333.77 336.05 -0.6689 ÉTAIENT TOMBÉS DANS UN OUBLI PROFOND
utt_0109 utt 336.71 338.59 -1.3516 PERSONNE N'EÛT OSÉ EN PARLER
utt_0110 utt 338.73 341.26 -1.1054 PERSONNE N'EÛT MÊME OSÉ S'EN SOUVENIR
utt_0111 utt 342.97 347.09 -0.1635 MONSIEUR MYRIEL ÉTAIT ARRIVÉ À DIGNE ACCOMPAGNÉ D'UNE VIEILLE FILLE
utt_0112 utt 347.09 348.64 -0.1329 MADEMOISELLE BAPTISTINE
utt_0113 utt 348.83 352.08 -0.2481 QUI ÉTAIT SA SOEUR ET QUI AVAIT DIX ANS DE MOINS QUE LUI
utt_0114 utt 353.27 358.38 -0.4945 ILS AVAIENT POUR TOUT DOMESTIQUE UNE SERVANTE DU MÊME ÂGE QUE MADEMOISELLE BAPTISTINE
utt_0115 utt 358.49 360.23 -0.2872 ET APPELÉE MADAME MAGLOIRE
utt_0116 utt 360.63 361.43 -0.2986 LAQUELLE
utt_0117 utt 361.69 364.74 -0.6303 APRÈS AVOIR ÉTÉ _LA SERVANTE DE MONSIEUR LE CURÉ_
utt_0118 utt 364.99 372.23 -0.2863 PRENAIT MAINTENANT LE DOUBLE TITRE DE FEMME DE CHAMBRE DE MADEMOISELLE ET FEMME DE CHARGE DE MONSEIGNEUR
utt_0119 utt 373.63 376.70 -0.3932 MADEMOISELLE BAPTISTINE ÉTAIT UNE PERSONNE LONGUE
utt_0120 utt 376.99 377.70 -0.0751 PÂLE
utt_0121 utt 378.03 378.66 -0.3890 MINCE
utt_0122 utt 378.99 379.60 -0.2301 DOUCE
utt_0123 utt 380.39 385.24 -0.6638 ELLE RÉALISAIT L'IDÉAL DE CE QU'EXPRIME LE MOT «RESPECTABLE»
utt_0124 utt 385.81 391.29 -0.4178 CAR IL SEMBLE QU'IL SOIT NÉCESSAIRE QU'UNE FEMME SOIT MÈRE POUR ÊTRE VÉNÉRABLE
utt_0125 utt 391.75 393.48 -0.3714 ELLE N'AVAIT JAMAIS ÉTÉ JOLIE
utt_0126 utt 393.53 394.60 -0.1253 TOUTE SA VIE
utt_0127 utt 394.69 397.88 -1.1759 QUI N'AVAIT ÉTÉ QU'UNE SUITE DE SAINTES OEUVRES
utt_0128 utt 398.07 402.56 -0.4144 AVAIT FINI PAR METTRE SUR ELLE UNE SORTE DE BLANCHEUR ET DE CLARTÉ
utt_0129 utt 403.15 403.66 -0.0811 ET
utt_0130 utt 403.66 404.41 -0.2271 EN VIEILLISSANT
utt_0131 utt 404.55 409.00 -0.5422 ELLE AVAIT GAGNÉ CE QU'ON POURRAIT APPELER LA BEAUTÉ DE LA BONTÉ
utt_0132 utt 410.45 414.64 -0.2964 CE QUI AVAIT ÉTÉ DE LA MAIGREUR DANS SA JEUNESSE ÉTAIT DEVENU
utt_0133 utt 414.64 415.63 -0.0525 DANS SA MATURITÉ
utt_0134 utt 415.83 417.24 -0.1909 DE LA TRANSPARENCE
utt_0135 utt 418.11 422.44 -0.5773 ET CETTE DIAPHANÉITÉ LAISSAIT VOIR L'ANGE
utt_0136 utt 423.39 426.82 -0.5336 C'ÉTAIT UNE ÂME PLUS ENCORE QUE CE N'ÉTAIT UNE VIERGE
utt_0137 utt 427.57 430.10 -0.2424 SA PERSONNE SEMBLAIT FAITE D'OMBRE
utt_0138 utt 430.43 433.56 -0.7741 À PEINE ASSEZ DE CORPS POUR QU'IL Y EÛT LÀ UN SEXE
utt_0139 utt 433.68 436.17 -0.0782 UN PEU DE MATIÈRE CONTENANT UNE LUEUR
utt_0140 utt 436.42 438.47 -0.5383 DE GRANDS YEUX TOUJOURS BAISSÉS
utt_0141 utt 438.52 441.53 -0.9977 UN PRÉTEXTE POUR QU'UNE ÂME RESTE SUR LA TERRE
utt_0142 utt 443.08 445.61 -0.0723 MADAME MAGLOIRE ÉTAIT UNE PETITE VIEILLE
utt_0143 utt 445.88 446.63 -0.1467 BLANCHE
utt_0144 utt 446.86 447.65 -0.8112 GRASSE
utt_0145 utt 447.90 448.71 -0.1460 REPLÈTE
utt_0146 utt 448.92 449.79 -0.6227 AFFAIRÉE
utt_0147 utt 450.00 451.49 -0.3090 TOUJOURS HALETANTE
utt_0148 utt 451.62 453.71 -0.5239 À CAUSE DE SON ACTIVITÉ D'ABORD
utt_0149 utt 453.94 456.09 -0.4499 ENSUITE À CAUSE D'UN ASTHME
utt_0150 utt 458.38 459.47 -0.0230 À SON ARRIVÉE
utt_0151 utt 459.58 469.93 -1.2239 ON INSTALLA MONSIEUR MYRIEL EN SON PALAIS ÉPISCOPAL AVEC LES HONNEURS VOULUS PAR LES DÉCRETS IMPÉRIAUX QUI CLASSENT L'ÉVÊQUE IMMÉDIATEMENT APRÈS LE MARÉCHAL DE CAMP
utt_0152 utt 470.84 474.33 -0.3690 LE MAIRE ET LE PRÉSIDENT LUI FIRENT LA PREMIÈRE VISITE
utt_0153 utt 474.78 479.51 -1.0417 ET LUI DE SON CÔTÉ FIT LA PREMIÈRE VISITE AU GÉNÉRAL ET AU PRÉFET
utt_0154 utt 480.34 481.91 -0.3336 L'INSTALLATION TERMINÉE
utt_0155 utt 482.18 484.47 -1.0307 LA VILLE ATTENDIT SON ÉVÊQUE À L'OEUVRE
```

</div></details>

Maybe the timings on your shorter audio file are more accurate when you pre-set the correct timing information?
For the French model:
```python
aligner = CTCSegmentation(asr_model, kaldi_style_text=False, samples_to_frames_ratio=320, time_stamps="fixed")
```

TParcollet commented:
@lumaku as a final step, would you mind doing a very short Google Colab that we could put on the website? I think that audio segmentation is a commonly requested feature ... Maybe explaining all the parameters, and how to do it for an English / French / Chinese file, for instance :) Maybe comment on the duration problem as well. It doesn't have to be 50min long, it could be 5min only :) The PR looks good to me, if tests pass, we can merge. @mravanelli

lingjzhu commented:
@lumaku @mravanelli Thanks very much for your detailed replies about CTC-based forced alignment! I have tested phoneme forced alignment using CTC outputs. I trained a wav2vec2 model fine-tuned on Librispeech. The results are OK because each phone is at least aligned to one correct frame. However, the phoneme boundaries are very inaccurate. On average, the forced aligned boundaries are about 20~60ms off the ground truth, which is consistent with your previous observations. Directly using CTC outputs might not necessarily be better than the traditional systems like Montreal Forced Aligner.

mravanelli commented:
Thanks for the feedback! Note that in SpeechBrain we already have some
systems for HMM-based aligment (see TIMIT/Alignments). It could be great to
see a comparison between these systems and the pure CTC object of this PR.

On Wed, Jul 21, 2021, 10:06 PM jzhu ***@***.***> wrote:

> @lumaku <https://github.com/lumaku> @mravanelli
> <https://github.com/mravanelli> Thanks very much for your detailed
> replies about CTC-based forced alignment! I have tested phoneme forced
> alignment using CTC outputs. I trained a wav2vec2 model fine-tuned on
> Librispeech. The results are OK because each phone is at least aligned to
> one correct frame. However, the phoneme boundaries are very inaccurate. On
> average, the forced aligned boundaries are about 20~60ms off the ground
> truth, which is consistent with your previous observations. Directly using
> CTC outputs might not necessarily be better than the traditional systems
> like Montreal Forced Aligner.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/speechbrain/speechbrain/pull/819#issuecomment-884610307>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEA2ZVRNPTDBEJNOFT2YB2LTY54KLANCNFSM46VEVJPQ>
> .
>

lumaku commented:
@lingjzhu @TParcollet  On other point regarding timing: I haven't thoroughly looked into the frontend yet; but the Speechbrain frontend has integrated padding at the input. Padding can shift timing estimates by a few ms...

I collected the examples above together in a short gist.
https://gist.github.com/lumaku/75eca1c86d9467a54888d149dc7b84f1

The use case for CTC segmentation is rather for larger files, and we performed alignment and clean-up of files of over 3h audio with it. This can be comfortably done with RNN-based models, but Transformer-based models do not scale to longer files. Speechbrain only has Transformer-based models available? As a solution, it's possible to partition longer audio files, I wrote some information about this in the gist.

TParcollet commented:
@lumaku sorry for the very very late reply on that part ..... Thank you very much for the gist, but SpeechBrain has standardised its tutorials to be on Colab so peoples can play with it directly. Do you mind converting it ?

Thanks !

TParcollet commented:
Let me merge it, but please @lumaku, let me know if you can provide a tutorial on Colab instead.

lumaku commented:
@TParcollet  Sorry for the delay, after the summer break I have too much tasks, so that I won't be able to write a Colab notebook in the near term. However, let me know if you have any questions regarding the tool.
